src/scope.ml

```diff
@@ -137,24 +162,36 @@ let scope : mode -> sig_state -> env -> p_term -> tbox = fun md ss env t ->
       _Patt None name (Env.to_tbox env)
   in
   (* Toplevel scoping function, with handling of implicit arguments. *)
-  let rec scope : env -> p_term -> tbox = fun env t ->
+  let rec scope : ?allow_private:bool -> env -> p_term -> tbox =
```

Comment:
Why do you take an optional argument here? Isn't it possible to use a boolean instead? The semantics would be clearer.

Reply:
It would be a boolean option: on the first scoping of a rewrite rule, this option is set to `Some(b)` to indicate whether private symbols are allowed. Therefore, the first run of `scope` is performed without indicating the argument.

Reply:
I don't understand. Could you please elaborate? Why are you talking of several scoping runs/passes? Note that, after lines 255-256, None seems to be equivalent to Some(true).

Reply:
When we handle a definition or a theorem, allow_private is given by the expostag of the definition/theorem: allow_private=true if expostag=Private, and false otherwise.
When we handle a tactic or a query, we can take allow_private=true.
When we handle a rule l->r, we could start by checking that l is of the form f l1 .. ln with f a function symbol, without calling scope. Then, scoping l1 .. ln with allow_private=true and r with allow_private=(f is Private).

Reply:
The first attempt was to get `f` itself to see its exposition, but it seems that because of infix symbols, this appeared tricky. What is done now is:
- scope `l -> r`
- the call to `scope_head` allows to retrieve the root symbol; so we allow private symbol to this first recursive call
- the rest of the rule is scoped with the exposition of the root symbol.

My "scoping runs" referred to the successive recursive calls of the `scope` function.

Reply:
I don't understand why infix symbols would be a problem. Looking more at the code, I feel that there is something strange concerning unary and binary symbol applications @rlepigre . Syntax.get_args does not really return the head term nor all the arguments. If the term is a P_BinO or P_UnaO, it does not decompose it. This is not documented. Now, P_UnaO((op,_,f),t) should be the same as P_Appl(Symb(f,Unary(op)),t), and similarly for P_BinO. Well, in fact, the parser could certainly replace one by the other so that P_UnaO and P_BinO could be removed from p_term. Why don't we do that?

Reply:
It would be a really bad idea to remove these constructors. They are useful for many good reason, for example to be able to print back the AST (although this is not yet done correctly). During scoping, unary and binary symbols should be desugared early enough so that they do not interfere with further computations.

Reply:
I don't understand why you want to be able to print out terms as they have been entered. This is not important to me. What is important is to print terms in a uniform and predictable way. This is related to #260.

Reply:
Because you want the user to be able to recognize their own terms! For example if there is a notation in scope that the user does not know about or does not want to use it is not a good idea to force them to see it when printing subterms of what the user typed.

Reply:
Let's continue the discussion on notations in #260.

Reply:
@rlepigre for what other reasons the constructors P_UnaO and P_BinO are useful?

Reply:
@gabrielhdt could you please explain why it would be tricky with infix symbols? Is it because Syntax.get_args is not correct? As this function is used only in scope, we could change it.

Reply:
Another reason why they are useful is to have correct positions in scoping.

Reply:
I don't understand. Could you elaborate?

Reply:
Well, in the current representation, if you have a term representing `e1 + e2` then you have the position of the whole expression and you also have access to the positions of `e1` and `e2`. On the other hand, if you build a term representing `(add e1) e2` from a parsed infix notation then you have no correct source code position for the subterm `add e1`, nor for `add` itself.

Reply:
Ok but, currently, this is not used much. Remark: the position of add could be set to that of +.

Reply:
@fblanqui it was precisely because of `Syntax.get_args`, I expected it to do like `Basics.get_args`, which for `3 + 2` would return `(+, [3; 2])` (at the level of `p_term`) but apparently it does not.

Reply:
Ok. What if you fix Syntax.get_args? Could you try to use a bool instead of a bool option in this case?

Reply:
Indeed I could, I would scope the root symbol alone, get its privacy and then scope the rest of the lhs with the boolean set accordingly.

Reply:
It would actually make sense to move `Syntax.get_args` inside the definition of the scoping function (it is only useful there anyway). That way you can actually move the code that unsugars the unary and binary symbols into the definition of `get_args`, and so the unary and binary symbol constructors are then eliminated early.